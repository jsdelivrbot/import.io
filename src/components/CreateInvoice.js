import React, { Component } from 'react';
import {connect} from 'react-redux';
import * as actions from '../actions';
import { Button } from 'react-bootstrap';



class CreateInvoice extends Component {

  constructor () {
    super();

    this.state = {
      item: '',
      quantity: '',
      price: '',
      formValItem: 'none',
      formValQuantity: 'none',
      formValPrice: 'none',
      invoiceId: 0
    };

    this.handleFormValidation = this.handleFormValidation.bind(this);

  }

  handleChangeItem (e) {
    this.setState({
      item: e.target.value
    });
  }

  handleChangeQuantity (e) {
    this.setState({
      quantity: e.target.value
    });
  }

  handleChangePrice (e) {
    this.setState({
      price: e.target.value
    });
  }


  handleFormValidation (item, quantity, price) {


    // reset UI warnings to give people the chance to start fresh on each edit
    this.setState({
      formValItem: 'none',
      formValQuantity: 'none',
      formValPrice: 'none'
    })


    if (item === '') {
      this.setState({
        formValItem: 'inline-block'
      })
      return false;
    } else {
      this.setState({
        formValItem: 'none'
      })
    }


    if (quantity.match(/^[0-9]+$/) && quantity !== '') {
      this.setState({
        formValQuantity: 'none'
      })
    } else {
      this.setState({
        formValQuantity: 'inline-block'
      })
      return false;
    }

    if ( price.match(/^[0-9.]+$/) && price !== '') {
      let decimalCount = 0;

      for (let i = 0; i < price.length; i++) {
        if (price[i] === '.') {
          decimalCount++;
        }

        if (decimalCount > 1 ) {
          this.setState({
            formValPrice: 'inline-block'
          })
          return false;
        } else {
          this.setState({
            formValPrice: 'none'
          })
        }
      }
    } else {
      this.setState({
        formValPrice: 'inline-block'
      })
      return false;
    }

    return true;
  }

  handleSubmit(item, quantity, price) {
    let id = this.state.invoiceId++

    if (this.handleFormValidation(item, quantity, price)) {

      this.props.createInvoice( id, item, quantity, price);

      this.setState({
        item: '',
        quantity: '',
        price: '',
      });
    }
  }


  render () {
    return (
      <thead className='table-header'>
        <tr>
          <th>
            <p style={{display: this.state.formValItem, color: 'red'}}>Please enter an item</p>
            <input style={{width: '200px'}} onChange={(e) => this.handleChangeItem(e)} type='text' placeholder='item' value={this.state.item} className='form-control'/>
          </th>
          <th>
            <p style={{display: this.state.formValQuantity, color: 'red'}}>Please enter a quantity</p>
            <input style={{width: '200px'}} onChange={(e) => this.handleChangeQuantity(e)} type='text' placeholder='quantity' value={this.state.quantity} className='form-control'/>
          </th>
          <th>
            <p style={{display: this.state.formValPrice, color: 'red'}}>Please enter a price</p>
            <input style={{width: '200px'}} onChange={(e) => this.handleChangePrice(e)} type='text' placeholder='price' value={this.state.price} className='form-control'/>
          </th>
          <th></th>
          <th></th>
          <th>
            <Button bsSize='small' bsStyle='primary' onClick={() => {this.handleSubmit( this.state.item, this.state.quantity, this.state.price)}}>Add Invoice</Button>
          </th>
        </tr>
      </thead>
    );
  }
}


export const Unwrapped = CreateInvoice
export default connect(null, actions)(CreateInvoice)
