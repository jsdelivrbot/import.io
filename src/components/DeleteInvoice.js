import React, { Component } from 'react';
import {connect} from 'react-redux';
import * as actions from '../actions';
import { Button } from 'react-bootstrap';


class DeleteInvoice extends Component {

  constructor () {
    super();

    this.handleFormSubmit = this.handleFormSubmit.bind(this);
  }


  handleFormSubmit(e) {
    if (e.target.value === 'submit') {
      this.props.deleteInvoice(this.props.invoice.id, this.props.invoice.quantity, this.props.invoice.price);
      this.props.hide();
    } else if (e.target.value === 'close') {
      this.props.hide();
    }
  }


  render () {
    return (
      <div className='delete-confirmation-modal'>
        <h6>Are you sure?</h6>
        <div>
          <Button onClick={(e) => this.handleFormSubmit(e)} value='close' className='btn btn-info btn-sm delete-button'>Cancel</Button>
          <Button onClick={(e) => this.handleFormSubmit(e)} value='submit' className='btn btn-danger btn-sm delete-button'>Delete</Button>
        </div>
      </div>
    );
  }
}


export const Unwrapped = DeleteInvoice
export default connect(null, actions)(DeleteInvoice)
