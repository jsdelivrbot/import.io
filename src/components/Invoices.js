import React, { Component } from 'react';
import { connect } from 'react-redux';
import * as actions from '../actions' ;
import CreateInvoice from './CreateInvoice';
import UpdateInvoiceButton from './UpdateInvoiceButton';
import DeleteInvoiceButton from './DeleteInvoiceButton';



class Invoices extends Component {
  constructor() {
    super();

    this.state = {
      updated: false
    };

    this.handleUpdating = this.handleUpdating.bind(this);
  }


  componentDidMount() {
    this.props.getInvoices();
  }

  handleUpdating() {
    this.setState({
      updated: !this.state.updated
    });
  }

  componentWillReceiveProps(nextProps) {
    let invoices = this.props.allInvoices;

    // this will only trigger on an Edit
    if (this.state.updated) {
      this.setState({
        updated: !this.state.updated
      });

      this.renderTable(nextProps.allInvoices);
    }

    // this will only trigger on a Create/Delete
    if ( invoices !== undefined && nextProps.allInvoices !== undefined ) {
      if (invoices.length !== nextProps.allInvoices.length) {
        this.props.getInvoices();
      }
    }
  }

  renderTable () {
    if (this.props.allInvoices) {
      return this.props.allInvoices.map((invoice) =>
          <tr key={invoice.id} >
            <td >{invoice.item}</td>
            <td>{invoice.quantity}</td>
            <td >${invoice.price}</td>
            <td>${invoice.quantity * invoice.price}</td>
            <td>
              <UpdateInvoiceButton invoice={invoice} handleUpdating={this.handleUpdating}/>
            </td>
            <td>
              <DeleteInvoiceButton invoice={invoice}/>
            </td>
          </tr>
      );
    }
  }



  render () {
    return (
      <div >
        <h1 className='invoices-header'>Invoices</h1>

        <div className='table-container'>
          <table className="table table-hover col-xs-2 ">
            <thead className='table-header'>
              <tr>
                <th>Item</th>
                <th>Qty</th>
                <th>Price</th>
                <th>Total</th>
                <th>Edit</th>
                <th>Delete</th>
              </tr>
            </thead>
            <tbody >
              {this.renderTable()}
            </tbody>
              <CreateInvoice />
            <tbody>
              <tr>
                <td></td>
                <td></td>
                <td></td>
                <td></td>
                <td>Subtotal</td>
                <td>${this.props.subtotal}</td>
              </tr>
              <tr>
                <td></td>
                <td></td>
                <td></td>
                <td></td>
                <td>Tax(5%)</td>
                <td>${this.props.tax}</td>
              </tr>
              <tr>
                <td></td>
                <td></td>
                <td></td>
                <td></td>
                <td>Total</td>
                <td>${this.props.total}</td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    );
  }
}


function mapStateToProps(state) {
  return {
    allInvoices: state.invoices.allInvoices,
    subtotal: state.invoices.subtotal,
    tax: state.invoices.tax,
    total: state.invoices.total
  };
}

export const Unwrapped = Invoices
export default connect(mapStateToProps, actions)(Invoices);


