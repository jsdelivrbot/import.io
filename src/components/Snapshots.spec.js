import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import { Unwrapped as UnwrappedInvoices } from './Invoices'
import { Unwrapped as UnwrappedDeleteInvoice } from './DeleteInvoice'
import { Unwrapped as UnwrappedCreateInvoice } from './CreateInvoice'
import { Unwrapped as UnwrappedUpdateInvoiceModalForm } from './CreateInvoice'


test('Invoices snapshot test', () => {
  const component = shallow(<UnwrappedInvoices />)
  const tree = shallowToJson(component)

  expect(tree).toMatchSnapshot()
})

test('CreateInvoice snapshot test', () => {
  const component = shallow(<UnwrappedCreateInvoice />)
  const tree = shallowToJson(component)

  expect(tree).toMatchSnapshot()
})

test('DeleteInvoice snapshot test', () => {
  const component = shallow(<UnwrappedDeleteInvoice />)
  const tree = shallowToJson(component)

  expect(tree).toMatchSnapshot()
})


test('UpdateInvoiceModalForm snapshot test', () => {
  const component = shallow(<UnwrappedUpdateInvoiceModalForm />)
  const tree = shallowToJson(component)

  expect(tree).toMatchSnapshot()
})