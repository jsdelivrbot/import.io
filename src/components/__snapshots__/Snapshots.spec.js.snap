exports[`test CreateInvoice snapshot test 1`] = `
<thead
  className="table-header">
  <tr>
    <th>
      <p
        style={
          Object {
            "color": "red",
            "display": "none",
          }
        }>
        Please enter an item
      </p>
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="item"
        style={
          Object {
            "width": "200px",
          }
        }
        type="text"
        value="" />
    </th>
    <th>
      <p
        style={
          Object {
            "color": "red",
            "display": "none",
          }
        }>
        Please enter a quantity
      </p>
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="quantity"
        style={
          Object {
            "width": "200px",
          }
        }
        type="text"
        value="" />
    </th>
    <th>
      <p
        style={
          Object {
            "color": "red",
            "display": "none",
          }
        }>
        Please enter a price
      </p>
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="price"
        style={
          Object {
            "width": "200px",
          }
        }
        type="text"
        value="" />
    </th>
    <th />
    <th />
    <th>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsSize="small"
        bsStyle="primary"
        disabled={false}
        onClick={[Function]}>
        Add Invoice
      </Button>
    </th>
  </tr>
</thead>
`;

exports[`test DeleteInvoice snapshot test 1`] = `
<div
  className="delete-confirmation-modal">
  <h6>
    Are you sure?
  </h6>
  <div>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="default"
      className="btn btn-info btn-sm delete-button"
      disabled={false}
      onClick={[Function]}
      value="close">
      Cancel
    </Button>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="default"
      className="btn btn-danger btn-sm delete-button"
      disabled={false}
      onClick={[Function]}
      value="submit">
      Delete
    </Button>
  </div>
</div>
`;

exports[`test Invoices snapshot test 1`] = `
<div>
  <h1
    className="invoices-header">
    Invoices
  </h1>
  <div
    className="table-container">
    <table
      className="table table-hover col-xs-2 ">
      <thead
        className="table-header">
        <tr>
          <th>
            Item
          </th>
          <th>
            Qty
          </th>
          <th>
            Price
          </th>
          <th>
            Total
          </th>
          <th>
            Edit
          </th>
          <th>
            Delete
          </th>
        </tr>
      </thead>
      <tbody />
      <Connect(CreateInvoice) />
      <tbody>
        <tr>
          <td />
          <td />
          <td />
          <td />
          <td>
            Subtotal
          </td>
          <td>
            $
          </td>
        </tr>
        <tr>
          <td />
          <td />
          <td />
          <td />
          <td>
            Tax(5%)
          </td>
          <td>
            $
          </td>
        </tr>
        <tr>
          <td />
          <td />
          <td />
          <td />
          <td>
            Total
          </td>
          <td>
            $
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`test UpdateInvoiceModalForm snapshot test 1`] = `
<thead
  className="table-header">
  <tr>
    <th>
      <p
        style={
          Object {
            "color": "red",
            "display": "none",
          }
        }>
        Please enter an item
      </p>
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="item"
        style={
          Object {
            "width": "200px",
          }
        }
        type="text"
        value="" />
    </th>
    <th>
      <p
        style={
          Object {
            "color": "red",
            "display": "none",
          }
        }>
        Please enter a quantity
      </p>
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="quantity"
        style={
          Object {
            "width": "200px",
          }
        }
        type="text"
        value="" />
    </th>
    <th>
      <p
        style={
          Object {
            "color": "red",
            "display": "none",
          }
        }>
        Please enter a price
      </p>
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="price"
        style={
          Object {
            "width": "200px",
          }
        }
        type="text"
        value="" />
    </th>
    <th />
    <th />
    <th>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsSize="small"
        bsStyle="primary"
        disabled={false}
        onClick={[Function]}>
        Add Invoice
      </Button>
    </th>
  </tr>
</thead>
`;
